### PR TITLE
Pretty printing removes array indices #29

### PR DESCRIPTION
**Pull request by Himeshi**

### CODE REVIEW

File: lib/Core/PrettyExpressionBuilder.cpp [hunk 1]
```diff
@@ -28,6 +28,9 @@ using namespace klee;
 
 namespace klee {
 
+const std::string array_prefix32 = "__arr32__";
+const std::string array_prefix64 = "__arr64__";
```
Comment (domainexpert) [inline lib/Core/PrettyExpressionBuilder.cpp]: @Himeshi It seems there are missing cases of arrays of 16 bits and 8 bits (byte) here.

**Review by domainexpert — CHANGES_REQUESTED** [review]
@Himeshi I think 16-bit and 8-bit cases are not handled.

**Review by domainexpert — APPROVED** [review]

**Comment by domainexpert** [comment]
@Himeshi Would you please rebase to master and please let me know how to test that this fix is working?

File: lib/Core/PrettyExpressionBuilder.cpp [hunk 2]
```diff
@@ -169,9 +174,40 @@ std::string PrettyExpressionBuilder::writeExpr(std::string array,
...
 std::string PrettyExpressionBuilder::readExpr(std::string array,
                                               std::string index) {
-  return array;
+  int array_index;
+  if (!array.compare(0, array_prefix8.size(), array_prefix8)) {
+    std::istringstream ss(index);
+    ss >> array_index;
+    std::ostringstream so;
+    so << array_index;
+    return array.erase(0, 9) + "[" + so.str() + "]";
```
Comment (domainexpert) [inline lib/Core/PrettyExpressionBuilder.cpp]: @Himeshi The erase range here should be (0, 8).

**Review by domainexpert — CHANGES_REQUESTED** [review]
@Himeshi Just a tiny thing: the rest looks good to me.

**Comment by Himeshi** [comment]
@domainexpert This pull request can be tested with the attached program. Without this code the array indices should be missing in the error expression in the `test000001.presicion_error` file.

[array_test2.pdf](https://github.com/fp-analysis/klee/files/1233260/array_test2.pdf)

**Review by domainexpert — APPROVED** [review]

**Comment by domainexpert** [comment]
@Himeshi Awesome work. Thank you!

**Pull request merged**